syne_tune/backend/sagemaker_backend/sagemaker_utils.py

```diff
@@ -284,3 +285,144 @@ def map_identifier_limited_length(
         assert 1 < rnd_digits < max_length
         postfix = random_string(rnd_digits)
         return name[:(max_length - rnd_digits)] + postfix
+
+
+def _s3_traverse_recursively(
+        s3_client, action, bucket: str, prefix: str) -> dict:
+    """
+    Traverses directory from root `prefix`. The function `action` is applied
+    to all objects encountered, the signature is
+
+        `action(s3_client, bucket, object_key)`
+
+    'action' returns None if successful, otherwise an error message.
+    We return a dict with 'num_action_calls', 'num_successful_action_calls',
+    'first_error_message' (the error message for the first failed `action` call
+    encountered).
+
+    :param s3_client: S3 client
+    :param action: See above
+    :param bucket: S3 bucket name
+    :param prefix: Prefix from where to traverse, must end with '/'
+    :return: See above
+    """
+    more_objects = True
+    continuation_kwargs = dict()
+    list_objects_kwargs = dict(Bucket=bucket,
+                               Prefix=prefix,
+                               Delimiter='/')
+    all_next_prefixes = []
+    num_action_calls = 0
+    num_successful_action_calls = 0
+    first_error_message = None
+    while more_objects:
```

Comment:
Optional but the code would probably be easier with paginator, something like:

```python  
paginator = s3.get_paginator('list_objects_v2')
pages = paginator.paginate(**list_objects_kwargs)
  
for page in pages:
    for object in page['Contents']:
        print(object['Key'])
        all_next_prefixes.append(...)
```

https://yellowdesert.consulting/2019/07/29/how-to-paginate-in-boto3-use-collections-instead/

Generally when you have this next token, boto provides those "paginator" that allows to get an iterator without having to bother about passing next token around.

Reply:
Thanks. Let's keep it as it is, since it works.